[14.0] Add module account_move_name_sequence

alexis-via opened the pull request:
This module restores the good old behavior where journal entry numbers were generated from a sequence configured on the journal.

It uses the same datamodel as previous odoo versions, to facilitate the migration to odoo v14.

It is an alternative to PR https://github.com/OCA/account-invoicing/pull/932/files but with a different approach: here, we just try to restore the same behavior as previous odoo versions, we don't try to add new (great) features.

I plan to add tests soon.

alexis-via committed:
Add module account_move_name_sequence

This module restores the good old behavior where journal entry numbers
were generated from a sequence configured on the journal.

alexis-via commented:
I added tests and Travis is green.
Ready for review!

jcdrubay commented on account_move_name_sequence/models/account_move.py:
```diff
@@ -0,0 +1,32 @@
+            lambda m: (m.name == "/" or not m.name)
+            and m.state == "posted"
+            and m.journal_id
+            and m.journal_id.sequence_id
```
If you don't set the sequence_id field as required on the journal, it means that you could have mixed implementation. Is that ok? 

Personally, I would prefer having a consistent implementation across all journals.

Maybe it could be handled with a configuration.

Or maybe a simple note in the Readme:

```
Remarks:

- This module does not enforce the use of the Sequence for all journals. In case the Sequence field is not set on the Journal, then the module will fallback on the native behavior. If you want to enforce, you only need to override the `account.journal` and set the field sequence_id as required.
```


Actually, it is kind of required as you automatically create the sequence, but some funny user could remove the value from the field.

alexis-via replied:
In the code of v13, the sequence_id field has required=True in the field definition, cf https://github.com/odoo/odoo/blob/13.0/addons/account/models/account.py#L733
But, in the form view, it's required="0" cf https://github.com/odoo/odoo/blob/13.0/addons/account/views/account_view.xml#L274 in order to leave the field empty and let the inherit of create() auto-create the sequence.
When the company is created after the installation of the module, the sequence will be auto-created for every journal. But if you install the module after the creation of the company (which will happen for all existing v14 deployment that wish to go away of the sequence mess/bugs on journal entries), then you will have to create the sequence, unless if we provide a post-install script.
I'm OK to set the field as required and write a post-install script. But it means that you will have a rupture in the chronology of sequence right after the installation in all your journals and unless if you reconfigure all the journal sequence straight away after module installation. Like you, I also prefer to have a consistent implementation across all journals.

alexis-via replied:
@jcdrubay The sequence is now required and I added a post install script. I did that in a separate commit, so that we can easily remove it if other reviewers think it's a bad idea.

jcdrubay reviewed (approved):
I think it is a nice addition.
After code review I only have one small suggestion for further documentation in the README.

alexis-via committed:
sequence is now required on journals

Add post-install script to create a sequence for all existing journals
Update README accordingly

luc-demeyer commented:
@alexis-via 
I took a different angle on this change in V14 for a recent customer project, cf. https://github.com/Noviat/noviat-apps/tree/14.0/account_journal_entry_numbering_scheme

I made the journal sequence configurable, hence no need for sequences.
The drawback for the approach in my module is that you need a bit of regex expertise to configure your numbering, which is an expertise that an accountant usually doesn't have. Hence the accountant needs to rely upon his Odoo integrator to configure the journal numbering.

We could consider merging both approaches into one single OCA module. The advantage of the regex approach is zero configuration once the initial setup is done, hence allowing both options could be the best of all worlds.

alexis-via commented:
@luc-demeyer I had a look at your module, it's interesting! It's good to have the choice between the 2 approaches, but I don't think it's a good idea to merge the 2 modules. I think it's better to have 2 small modules with different approaches and give the choice.

alexis-via commented:
With my last commit that sets the sequence_id field as required and provides a post-install script to update all the journals, Travis fails on the error log "Table 'account_journal': unable to set NOT NULL on column 'sequence_id'". I'm surprised that such a log message makes Travis fail... I don't know how to solve this. Any idea ?

rafaelbn commented:
@alexis-via Alexis is this intentional? Sequence increase "uncontrolled" if you back to draft and confirm again for any modification.

![GIF 09-09-2021 1-44-12](https://user-images.githubusercontent.com/8736623/132599916-403dcad2-f856-47a4-adc9-0924504dcb9c.gif)

luc-demeyer commented:
@alexis-via 
An advantage to put both options in the same module is to allow a choice per journal.
Making the sequence_id a required field makes both modules incompatible with each other.

alexis-via commented:
@rafaelbn oups, sorry about that. It is now fixed. Thanks for catching this!

kittiu commented:
> @alexis-via
> An advantage to put both options in the same module is to allow a choice per journal.
> Making the sequence_id a required field makes both modules incompatible with each other.

IMHO, I would prefer to keep them separated. I find that, one module should do one thing and do it well, linux style :).
Having many different approach can confuse user.

kittiu commented:
@alexis-via in v13, thank you also for bringing back the v13 ways.

In the payment document was using separated sequence (than journal's) for Customer Invoice (CUST.IN) and Vendor Bill (SUPP.OUT), for same Journal (Bank)

![Selection_129](https://user-images.githubusercontent.com/1973598/132803348-ecc6a8e7-7b9c-4b16-9c71-1215eab02db0.png)

![Selection_130](https://user-images.githubusercontent.com/1973598/132803362-4dc76ddc-1400-449e-a7f3-3b498603aee8.png)

Which I think it is the correct way for the Payment Document. Do you think this module could as do the same in v14 too?

![Selection_128](https://user-images.githubusercontent.com/1973598/132803663-d5fa4767-8910-4c6c-b3fb-1476edba56f5.png)

WDYT?

alexis-via commented:
@kittiu No, I don't plan to change anything on account.payment. In v14, account.payment inherit from account.move, so the name of account.payment is in fact the "name" of account.move. It has been the same for a long time on invoices: the invoice number is the number of the related account.move (for exemple, see the definition of the "number" field of account.invoice on v8 https://github.com/odoo/odoo/blob/8.0/addons/account/account_invoice.py#L208). I don't see any drawback on this point. Why do you think it may be a problem ?

kittiu commented:
@alexis-via in v8, this is the problem in our accounting practice. Most (not all) of customers, want to run payment document based on payment type (out or in). That's why v13 was correct for us. May be it is on country basis, so no problem.

alexis-via commented:
@kittiu I'm not sure I understand your point: you want a different sequence for incoming payment and outgoing payment, is that right ? Something similar to the double sequence for sale/purchase journal: 1 sequence for invoice and 1 sequence for refunds?

kittiu commented:
> @kittiu I'm not sure I understand your point: you want a different sequence for incoming payment and outgoing payment, is that right ? Something similar to the double sequence for sale/purchase journal: 1 sequence for invoice and 1 sequence for refunds?

Yes, that can be similar.
In v13, it was a fixed sequence for incoming and outgoing payment. But should you make it configurable in Journal setting, similar to credit note, should answer our case too. WDYT?

alexis-via commented:
I don't plan to change anything for account.payment because I prefer to have the move name used as account.payment name, to keep this module as simple as possible. But you could develop another small module to have a dedicated sequence for account.payment.

remi-filament commented on account_move_name_sequence/views/account_journal.xml:
```diff
@@ -0,0 +1,32 @@
+                                          <field
+                    name="refund_sequence"
+                    attrs="{'invisible': [('type', 'not in', ('sale', 'purchase'))]}"
+                />
```
This field is already present by default in journal view in Odoo (https://github.com/odoo/odoo/blob/755c074a1e4a17e3947d1017446e0d58e2ecacb3/addons/account/views/account_journal_views.xml#L69) , should be removed from here.

alexis-via replied:
fixed, thanks for spotting this

remi-filament commented:
I have tested more thoroughly this module and I face an issue when the sequence is not as expected by Odoo.
For instance if I set my sequence with : 
`FAC/%(y)s%(month)s/`

I get the following error 
```
The Date (27/10/2021) doesn't match the Numéro (FAC/2110/0001).
You might want to clear the field Numéro before proceeding with the change of the date.
```

This is caused by the constraint added by sequence_mixin. These checks should be disabled in case there is a sequence defined on journal I guess, as done here : https://github.com/OCA/account-invoicing/pull/932/files#diff-0eb1d66c80f60fa1d7232a235932ce7851b110afd77ddb7e730206fe65c7307cR55-R59

alexis-via commented:
@remi-filament fixed, thanks for spotting the bug.

remi-filament reviewed (approved):
Works fine now, thanks @alexis-via

clementmbr commented:
Hi @alexis-via @fmdl @dreispt,
I'm just checking this PR and I'm wondering if this PR only needs a rebase to pass Travis and be merged or if there's still something blocking it.

Thanks!

dreispt commented:
I see a red build, so that's blocking.
There is a chance a rebase can fix that.

rafaelbn commented:
/ocabot rebase

rafaelbn commented:
@alexis-via please could you please rebase in order to merge this one? Thank you!

Shide commented on account_move_name_sequence/views/account_journal.xml:
```diff
@@ -0,0 +1,30 @@
+                <field
+                    name="sequence_id"
+                    required="0"
+                    context="{'default_name': name, 'default_company_id': company_id, 'default_implementation': 'no_gap', 'default_padding': 4, 'default_use_date_range': True, 'default_prefix': code + '/%%(range_year)s/'}"
```
If you click on `sequence_id` before setting up `code`, throws an error (Unable to join bool (False) + str)

I suggest:
1. Fallback with a default text: `'default_prefix': (code or 'UNKNOWN') + '/%%(range_year)s/'`
2. Invisible by attrs: `attrs="{'invisible': [('code', '=', False)]}"`

As you like :)

Shide commented on account_move_name_sequence/views/account_journal.xml:
```diff
@@ -0,0 +1,30 @@
+                <field
+                    name="refund_sequence_id"
+                    attrs="{'invisible': ['|', ('type', 'not in', ('sale', 'purchase')), ('refund_sequence', '=', False)]}"
+                    context="{'default_name': name, 'default_company_id': company_id, 'default_implementation': 'no_gap', 'default_padding': 4, 'default_use_date_range': True, 'default_prefix': 'R' + code + '/%%(range_year)s/'}"
```
Same as above

Shide reviewed (changes requested):
Super minor fixes. GJ!

rafaelbn commented:
Hello @alexis-via , could you please review https://github.com/oca/account-financial-tools/pull/1228#pullrequestreview-920474078 commnet and make a rebase to see if bots are happy? Now they are 🔴 

> I plan to add tests soon.

This would be great!

alexis-via commented:
Making Travis green is not easy, because I add some fields to account.journal with required=True, and the demo DB already has some journals. I don't know what would be the best strategy to make Travis green for this module ; any help would be appreciated.

dreispt commented:
@alexis-via Note that if you add a required field to a table with data, it won't be craeted since Postgres can't enforce it.
Some strategies to work around this are:
a) don't make it required at the database level, only at the UI level, or
b) don't make it required at the database level, but add a Python check constraint making sure it is set for new and edited records.
c) have it required at the db level, but add a pre-init hook that created the database column and populates it with a default values.

moylop260 commented:
@alexis-via 

I have added new code to this module in order to create the sequence with current moves if you have a database already used in production

And compute the current next number for each date range

I can take care of the travis red too

Do you think I should create a new PR to OCA or should I create a PR to your PR?

Let me know the best way for you, please

BTW thank you for take care about this feature/module!


UPDATED: PR to PR created here
 - https://github.com/akretion/account-financial-tools/pull/8/files

rafaelbn commented:
If you agree @alexis-via I think the best option could be that @moylop260 make a PR to your PR, ni any case, could you please answer to the lovely proposition 😃 

>Do you think I should create a new PR to OCA or should I create a PR to your PR?

> I can take care of the travis red too

Really kind @moylop260 ! 😄 Thank you! ❤️

moylop260 commented:
I have created the following PR https://github.com/OCA/account-financial-tools/pull/1394

in order to check if I'm already fixing the CI pipelines

Please, review

I even have the following PR created to this PR:
 - https://github.com/akretion/account-financial-tools/pull/8/files

remi-filament commented:
Thanks @moylop260 your PR makes sense.
I am however wondering if you could check in post_install.py whether these sequences already exist before creating new ones (for instance when migrating from previous version, we would already have defined sequence that we want to reuse instead of creating new ones)

moylop260 commented:
@remi-filament 

I have fixed your review here:
 - https://github.com/OCA/account-financial-tools/pull/1394

moylop260 commented:
FYI the PR is already green in the pipelines of github:
 - https://github.com/OCA/account-financial-tools/pull/1394

remi-filament commented:
Thanks @moylop260 !

moylop260 commented:
Superseded by https://github.com/OCA/account-financial-tools/pull/1394
